File: core/evidence/epistemic_evidence_engine.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

from core.evidence.epistemic_confidence_engine import score_epistemic_confidence
from core.evidence.evidence_sufficiency_engine import assess_evidence_sufficiency
from core.evidence.incompleteness_engine import preserve_incompleteness
from core.evidence.insufficiency_engine import mark_insufficiency
from core.evidence.semantic_reliability_engine import score_reliability
from core.evidence.semantic_support_engine import build_support
from core.evidence.semantic_uncertainty_propagation_engine import propagate_uncertainty
from core.evidence.cognitive_integrity_engine import apply_cognitive_integrity
from core.evidence.semantic_restraint_engine import apply_epistemic_restraint
from core.evidence.cognitive_humility_engine import apply_cognitive_humility
from core.evidence.reality_alignment_engine import apply_reality_alignment
from core.evidence.truth_preservation_engine import apply_truth_preservation
from core.evidence.recursive_reality_integrity_engine import apply_recursive_reality_integrity
from core.evidence.epistemic_civilization_stability_engine import apply_epistemic_civilization_stability
from core.evidence.cognitive_anti_capture_engine import apply_cognitive_anti_capture
from core.evidence.recursive_epistemic_sovereignty_engine import apply_recursive_epistemic_sovereignty
from core.evidence.civilizational_epistemic_openness_engine import apply_civilizational_epistemic_openness
from core.semantic.contradiction_restraint_engine import apply_contradiction_restraint
from core.evidence.semantic_weakness_engine import build_weaknesses


def attach_epistemic_state(bundle: Dict[str, Any]) -> Dict[str, Any]:
    evidence: List[str] = list(bundle.get("evidence", []) or [])
    ambiguities: List[str] = list(bundle.get("ambiguities", []) or [])
    contradicted = bundle.get("contradicted", {}) or {}
    pairs = contradicted.get("pairs", []) if isinstance(contradicted, dict) else []
    contradicting = [f"contradiction:{p}" for p in pairs]
    parser_basis = bundle.get("parser_basis", {}) or {}
    parser_density = int(parser_basis.get("symbol_count", 0) or 0)

    support = build_support(evidence)
    weaknesses = build_weaknesses(evidence, ambiguities)
    sufficiency = assess_evidence_sufficiency(evidence)
    reliability = score_reliability(evidence, ambiguities, len(pairs))
    incompleteness = preserve_incompleteness(bundle)
    insufficiency = mark_insufficiency(bundle)
    epistemic_confidence = score_epistemic_confidence(
        evidence=evidence,
        contradicting=contradicting,
        uncertainty_factors=ambiguities,
        parser_density=parser_density,
    )
    lineage = bundle.get("lineage", {})
    how = {"stages": lineage.get("stages", []), "depth": lineage.get("depth", 0)}

    bundle["support"] = support
    bundle["weaknesses"] = weaknesses
    bundle["evidence_sufficiency"] = sufficiency
    bundle["reliability"] = reliability
    bundle["epistemic_state"] = {
        **incompleteness,
        "insufficient": insufficiency["insufficient"],
        "confidence": epistemic_confidence["score"],
    }
    bundle["confidence_basis"] = epistemic_confidence
    bundle["how"] = how
    bundle = propagate_uncertainty(bundle)
    if insufficiency["insufficient"]:
        ambiguities = sorted(set(ambiguities + ["insufficient_evidence"]))
        bundle["ambiguities"] = ambiguities
    bundle = apply_cognitive_integrity(bundle)
    bundle = apply_contradiction_restraint(bundle)
    chain = apply_reality_alignment(apply_cognitive_humility(apply_epistemic_restraint(bundle)))
    return apply_civilizational_epistemic_openness(
        apply_recursive_epistemic_sovereignty(
            apply_cognitive_anti_capture(
                apply_epistemic_civilization_stability(
                    apply_recursive_reality_integrity(apply_truth_preservation(chain))
                )
            )
        )
    )

File: src/evidence/epistemicEvidenceEngine.ts
/**
 * Converted from Python: core/evidence/epistemic_evidence_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { scoreEpistemicConfidence } from "./epistemicConfidenceEngine.js";
import { assessEvidenceSufficiency } from "./evidenceSufficiencyEngine.js";
import { preserveIncompleteness } from "./incompletenessEngine.js";
import { markInsufficiency } from "./insufficiencyEngine.js";
import { scoreReliability } from "./semanticReliabilityEngine.js";
import { buildSupport } from "./semanticSupportEngine.js";
import { propagateUncertainty } from "./semanticUncertaintyPropagationEngine.js";
import { applyCognitiveIntegrity } from "./cognitiveIntegrityEngine.js";
import { applyEpistemicRestraint } from "./semanticRestraintEngine.js";
import { applyCognitiveHumility } from "./cognitiveHumilityEngine.js";
import { applyRealityAlignment } from "./realityAlignmentEngine.js";
import { applyTruthPreservation } from "./truthPreservationEngine.js";
import { applyRecursiveRealityIntegrity } from "./recursiveRealityIntegrityEngine.js";
import { applyEpistemicCivilizationStability } from "./epistemicCivilizationStabilityEngine.js";
import { applyCognitiveAntiCapture } from "./cognitiveAntiCaptureEngine.js";
import { applyRecursiveEpistemicSovereignty } from "./recursiveEpistemicSovereigntyEngine.js";
import { applyCivilizationalEpistemicOpenness } from "./civilizationalEpistemicOpennessEngine.js";
import { applyContradictionRestraint } from "../semantic/contradictionRestraintEngine.js";
import { buildWeaknesses } from "./semanticWeaknessEngine.js";

export function attachEpistemicState(bundle: any): any {
  var evidence: any = [...py.iter(py.or2(py.get(bundle, "evidence", []), () => ([])))];
  var ambiguities: any = [...py.iter(py.or2(py.get(bundle, "ambiguities", []), () => ([])))];
  var contradicted: any = py.or2(py.get(bundle, "contradicted", {}), () => ({}));
  var pairs: any = (((contradicted !== null && typeof contradicted === "object" && !Array.isArray(contradicted) && !(contradicted instanceof Set) && !(contradicted instanceof Map))) ? py.get(contradicted, "pairs", []) : []);
  var contradicting: any = py.iter(pairs).map((p: any) => `contradiction:${py.toStr(p)}`);
  var parser_basis: any = py.or2(py.get(bundle, "parser_basis", {}), () => ({}));
  var parser_density: any = py.toInt(py.or2(py.get(parser_basis, "symbol_count", 0), () => (0)));
  var support: any = buildSupport(evidence);
  var weaknesses: any = buildWeaknesses(evidence, ambiguities);
  var sufficiency: any = assessEvidenceSufficiency(evidence);
  var reliability: any = scoreReliability(evidence, ambiguities, py.len(pairs));
  var incompleteness: any = preserveIncompleteness(bundle);
  var insufficiency: any = markInsufficiency(bundle);
  var epistemic_confidence: any = scoreEpistemicConfidence(evidence, undefined, contradicting, ambiguities, parser_density);
  var lineage: any = py.get(bundle, "lineage", {});
  var how: any = {"stages": py.get(lineage, "stages", []), "depth": py.get(lineage, "depth", 0)};
  py.setItem(bundle, "support", support);
  py.setItem(bundle, "weaknesses", weaknesses);
  py.setItem(bundle, "evidence_sufficiency", sufficiency);
  py.setItem(bundle, "reliability", reliability);
  py.setItem(bundle, "epistemic_state", {...(incompleteness), "insufficient": py.at(insufficiency, "insufficient"), "confidence": py.at(epistemic_confidence, "score")});
  py.setItem(bundle, "confidence_basis", epistemic_confidence);
  py.setItem(bundle, "how", how);
  bundle = propagateUncertainty(bundle);
  if (py.truthy(py.at(insufficiency, "insufficient"))) {
    ambiguities = py.sorted(py.toSet(py.add(ambiguities, ["insufficient_evidence"])));
    py.setItem(bundle, "ambiguities", ambiguities);
  }
  bundle = applyCognitiveIntegrity(bundle);
  bundle = applyContradictionRestraint(bundle);
  var chain: any = applyRealityAlignment(applyCognitiveHumility(applyEpistemicRestraint(bundle)));
  return applyCivilizationalEpistemicOpenness(applyRecursiveEpistemicSovereignty(applyCognitiveAntiCapture(applyEpistemicCivilizationStability(applyRecursiveRealityIntegrity(applyTruthPreservation(chain))))));
}
export { applyCivilizationalEpistemicOpenness, applyCognitiveAntiCapture, applyCognitiveHumility, applyCognitiveIntegrity, applyContradictionRestraint, applyEpistemicCivilizationStability, applyEpistemicRestraint, applyRealityAlignment, applyRecursiveEpistemicSovereignty, applyRecursiveRealityIntegrity, applyTruthPreservation, assessEvidenceSufficiency, buildSupport, buildWeaknesses, markInsufficiency, preserveIncompleteness, propagateUncertainty, scoreEpistemicConfidence, scoreReliability };
